mqtt/mqtt-bridge/src/persist/waking_state/ring_buffer/mod.rs

```diff
@@ -353,248 +427,293 @@ impl RingBuffer {
 
 impl Drop for RingBuffer {
     fn drop(&mut self) {
-        let result = self.mmap.flush();
+        let result = self.file.flush();
         if let Some(err) = result.err() {
             error!("Failed to flush {:?}", err);
         }
+        let result = self.metadata_file.flush();
+        if let Some(err) = result.err() {
+            error!("Failed to flush metadata {:?}", err);
+        }
     }
 }
 
-fn create_file(file_path: &Path) -> IOResult<File> {
+fn create_file(file_path: &PathBuf) -> IOResult<File> {
+    if let Some(parent) = file_path.parent() {
+        create_dir_all(parent)?;
+    }
     OpenOptions::new()
         .read(true)
         .write(true)
         .create(true)
         .open(file_path)
 }
 
+fn save_ring_buffer_metadata(metadata: &RingBufferMetadata, file: &mut File) -> BincodeResult<()> {
+    let buf = bincode::serialize(metadata)?;
+    file.seek(SeekFrom::Start(0))?;
+    file.write_all(&buf)?;
+    file.flush()?;
+    Ok(())
+}
+
+fn retrieve_ring_buffer_metadata(file: &mut File) -> BincodeResult<RingBufferMetadata> {
+    let mut buf = vec![];
+    file.read_to_end(&mut buf)?;
+    if buf.is_empty() {
```

Comment:
If empty we will just have an empty/default metadata, which is fine as it means we can scan block at pos 0. As this should only be empty if we haven't had a successful operation yet.